### PR TITLE
image,seed/seedwriter: integrate manifest functionality into seed.Writer (3/3)

### PR DESCRIPTION
**Pull request by Meulengracht**
This is part 3 of the SeedManifest migration to seedwriter.Writer. It handles the actual integration of the manifest into the seedwriter.Writer instead of being image's responsibility.

**Parts**
1: https://github.com/snapcore/snapd/pull/12699 (move to seedwriter package)
2: https://github.com/snapcore/snapd/pull/12732 (rename to Manifest)
3: https://github.com/snapcore/snapd/pull/12700 (integration into seedwriter)
4: https://github.com/snapcore/snapd/pull/12733 (extend usage of manifest in seedwriter)

### CODE REVIEW

File: cmd/snap/cmd_prepare_image_test.go [hunk 1]
```diff
@@ -185,10 +186,10 @@ func (s *SnapPrepareImageSuite) TestReadSeedManifest(c *C) {
 	defer r()
```
Comment (ernestl) [inline cmd/snap/cmd_prepare_image_test.go:186]: Consider renaming: `TestReadSeedManifest`

File: seed/seedwriter/manifest_test.go [hunk 2]
```diff
@@ -55,8 +55,8 @@ func (s *manifestSuite) writeSeedManifest(c *C, contents string) string {
 	return manifestFile
 }
```
Comment (ernestl) [inline seed/seedwriter/manifest_test.go:57]: Did a scan of the file (`grep -rin "SeedManifest"`), and picked up the following for consideration: 

```
manifest_test.go:51:func (s *manifestSuite) writeSeedManifest(c *C, contents string) string {
manifest_test.go:63:func (s *manifestSuite) TestReadSeedManifestFullHappy(c *C) {
manifest_test.go:66:	manifestFile := s.writeSeedManifest(c, `# test line should not match
manifest_test.go:87:func (s *manifestSuite) TestReadSeedManifestParseFails(c *C) {
manifest_test.go:105:		manifestFile := s.writeSeedManifest(c, t.contents)
manifest_test.go:111:func (s *manifestSuite) TestReadSeedManifestNoFile(c *C) {
manifest_test.go:118:func (s *manifestSuite) testWriteSeedManifest(c *C, revisions map[string]*seedwriter.ManifestSnapRevision, vss map[string]*seedwriter.ManifestValidationSet) string {
manifest_test.go:126:func (s *manifestSuite) TestWriteSeedManifestNoFile(c *C) {
manifest_test.go:127:	filePath := s.testWriteSeedManifest(c, nil, nil)
manifest_test.go:131:func (s *manifestSuite) TestWriteSeedManifest(c *C) {
manifest_test.go:132:	filePath := s.testWriteSeedManifest(c,
manifest_test.go:160:func (s *manifestSuite) TestSeedManifestSetAllowedSnapRevisionInvalidRevision(c *C) {
manifest_test.go:166:func (s *manifestSuite) TestSeedManifestSetAllowedSnapRevision(c *C) {
manifest_test.go:175:func (s *manifestSuite) TestSeedManifestSetAllowedSnapRevisionTwice(c *C) {
manifest_test.go:188:func (s *manifestSuite) TestSeedManifestMarkSnapRevisionSeededWithAllowedHappy(c *C) {
manifest_test.go:204:func (s *manifestSuite) TestSeedManifestMarkSnapRevisionSeededNoMatchingAllowed(c *C) {
manifest_test.go:217:func (s *manifestSuite) TestSeedManifestMarkSnapRevisionSeededTwice(c *C) {
manifest_test.go:225:func (s *manifestSuite) TestSeedManifestMarkSnapRevisionSeededWrongRevision(c *C) {
manifest_test.go:233:func (s *manifestSuite) TestSeedManifestSetAllowedValidationSet(c *C) {
manifest_test.go:247:func (s *manifestSuite) TestSeedManifestSetAllowedValidationSetTwice(c *C) {
manifest_test.go:260:func (s *manifestSuite) TestSeedManifestSetAllowedValidationSetInvalidSequence(c *C) {
manifest_test.go:296:func (s *manifestSuite) TestSeedManifestMarkValidationSetSeededUsedHappy(c *C) {
manifest_test.go:364:func (s *manifestSuite) TestSeedManifestMarkValidationSetWeirdCases(c *C) {
manifest_test.go:383:func (s *manifestSuite) TestSeedManifestMarkValidationSetSeededUsedTwice(c *C) {
manifest_test.go:393:func (s *manifestSuite) TestSeedManifestMarkValidationSetSeededWrongSequence(c *C) {
manifest_test.go:403:func (s *manifestSuite) TestSeedManifestMarkValidationSetSeededWrongPinned(c *C) {
```

Comment (Meulengracht) [reply seed/seedwriter/manifest_test.go:57]: all of these are fixed in https://github.com/snapcore/snapd/pull/12732/commits/14e31c67df011fffa68e408d0c7730c373cb10f8

File: seed/seedwriter/writer.go [hunk 3]
```diff
@@ -46,6 +46,14 @@ type Options struct {
 	// TestSkipCopyUnverifiedModel is set to support naive tests
 	// using an unverified model, the resulting image is broken
 	TestSkipCopyUnverifiedModel bool
+
+	// Manifest is used to track snaps and validation sets that have
+	// been seeded. It can be pre-provided to provide specific revisions
+	// and validation-sets sequences.
```
Comment (ernestl) [inline seed/seedwriter/writer.go:52]: sets or set?

File: seed/seedwriter/writer.go [hunk 4]
```diff
@@ -199,6 +207,8 @@ type Writer struct {
 	systemSnap                      *SeedSnap
 	kernelSnap                      *SeedSnap
 	noKernelSnap                    bool
+	manifest                        *Manifest
```
Comment (ernestl) [inline seed/seedwriter/writer.go:210]: as discussed, looks like duplication of options

Comment (Meulengracht) [reply seed/seedwriter/writer.go]: Nice catch, reused ManifestPath from options instead. I do keep a seperate 'copy' of the Manifest as the manifest may be initialized to a new instance.

File: seed/seedwriter/writer.go (hunk 3 above)
Comment (Meulengracht) [reply seed/seedwriter/writer.go]: Good catch! Thanks!

**Review by ernestl — APPROVED** [review]
Looks good

File: seed/seedwriter/writer.go [hunk 5]
```diff
@@ -286,6 +302,13 @@ func New(model *asserts.Model, opts *Options) (*Writer, error) {
 	return w, nil
 }
 
+func initManifestFromOptions(opts *Options) *Manifest {
+	if opts.Manifest == nil {
+		return NewManifest()
+	}
+	return opts.Manifest
+}
```
Comment (pedronis) [inline seed/seedwriter/writer.go:310]: this could be a private method on Options

File: seed/seedwriter/writer.go [hunk 6]
```diff
@@ -1334,6 +1367,12 @@ func (w *Writer) WriteMeta() error {
 		return err
 	}
 
+	if w.opts.ManifestPath != "" {
+		if err := w.manifest.Write(w.opts.ManifestPath); err != nil {
+			return err
+		}
+	}
```
Comment (pedronis) [inline seed/seedwriter/writer.go:1374]: same

**Review by pedronis — COMMENTED** [review]
did a pass

File: seed/seedwriter/writer.go (hunk 5 above)
Comment (Meulengracht) [reply seed/seedwriter/writer.go]: Like this suggestion, done!

File: seed/seedwriter/writer.go [hunk 7]
```diff
@@ -1314,6 +1342,11 @@ func (w *Writer) SeedSnaps(copySnap func(name, src, dst string) error) error {
 				// record final destination path
 				sn.Path = dst
 			}
+			if !info.Revision.Unset() {
+				if err := w.manifest.MarkSnapRevisionSeeded(sn.Info.SnapName(), sn.Info.Revision); err != nil {
+					return fmt.Errorf("cannot record snap for manifest: %s", err)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply seed/seedwriter/writer.go:1349]: Good point, added unit test for this

**Review by ernestl — APPROVED** [review]

File: seed/seedwriter/writer.go [hunk 8]
```diff
@@ -56,6 +56,15 @@ type Options struct {
 	ManifestPath string
 }
 
+// initManifest returns either the manifest already provided by the
+// options, or if not provided, returns a newly initialized manifest.
+func (opts *Options) initManifest() *Manifest {
```
Comment (pedronis) [inline seed/seedwriter/writer.go:61]: this could just be called manifest() here

Comment (Meulengracht) [reply seed/seedwriter/writer.go]: yes I was actually doubting this one, so appreciate your input here!

**Pull request merged**